crates/evm/core/src/backend/code_cache.rs

```diff
@@ -0,0 +1,150 @@
+use alloy_chains::Chain;
+use alloy_provider::{Network, Provider};
+use alloy_transport::{Transport, TransportResult};
+use quick_cache::sync::Cache;
+use revm::primitives::{Address, Bytes};
+
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+struct CodeDetected {
+    /// The block number at which the code was detected.
+    block_number: u64,
+    /// The code at that block.
+    code: Bytes,
+}
+
+#[derive(Debug, Default, Clone)]
+struct CodeHistory {
```

Comment:
Why do we need the history?

Reply:
I updated the comments and types to be more clear, but it's the metadata for a cache entry, needed to track when the earliest we detected code was, and when the latest we didn't detect code was. 
If you are lte the no_code_detected_block_number then the address didn't have code. 
if you are gte the code_detected.block_number then return the immutable code.
If there is a cache entry and the requested block is between this range then you must fetch a new result.